src/Features/Core/Portable/GenerateMember/AbstractGenerateMemberService.cs

```diff
@@ -154,6 +154,17 @@ protected AbstractGenerateMemberService()
                 isStatic = false;
                 return;
             }
+            else if (syntaxFacts.IsNameOfSubpatternElement(expression) &&
+                     syntaxFacts.IsPropertySubpattern(expression.Parent.Parent.Parent))
+            {
+                // something like: { [|X|]: int i } or like: Blah { [|X|]: int i }
+                var propertySubpattern = expression.Parent.Parent.Parent;
+                var inferenceService = document.Document.GetLanguageService<ITypeInferenceService>();
+                typeToGenerateIn = inferenceService.InferType(semanticModel, propertySubpattern, objectAsDefault: true, cancellationToken) as INamedTypeSymbol;
```

Comment:
Note: I'll look at the `InferType` logic, but this should largely rely on `GetTypeInfo(propertyPattern).ConvertedType`.

Reply:
Ok.  that didn't exist at the time.  But if it does now, i can try to move to that.

Reply:
Cool. Ping me after that. I assume the type inferrer changes will be reverted (or much simplified) as a result.

Reply:
@CyrusNajmabadi Why is using this type inference API still needed? You should be able to get the containing pattern (parent of propertySubpattern) and get `.ConvertedType` off of that directly, just as @jcouv did in the completion work. Can you please explain what the benefit of doing this is?

Reply:
Is there any use of `InferType` for patterns anymore now that you can get `Type` and `ConvertedType` for patterns?

Reply:
Yes.  InferType will still be the codepath used for lots of other scenarios. For example, if i do:  ```{ X: Foo() }``` and we 'generate method' for 'Foo', it will use the type-inferred to try to figure out what type it has there.

Reply:
> @CyrusNajmabadi Why is using this type inference API still needed? You should be able to get the containing pattern (parent of propertySubpattern) and get .ConvertedType off of that directly, just as @jcouv did in the completion work. Can you please explain what the benefit of doing this is?

I coudl do that.  But it would duplicate the logic from TypeInferrer.  So i kinda like having it in one place.  I'm amenable to duplicating it though.

Reply:
I don't understand. `GetTypeInfo().ConvertedType` will give you exactly that type you want to generate into. @jcouv used exactly that to find out all the properties on this type in order to show completion for them. Can you show some scenarioes where this "inferred type" and the type returned by `ConvertedType` would be differnt?

Reply:
> I coudl do that. But it would duplicate the logic from TypeInferrer. 

I don't really see `pattern.GetTypeInfo().ConvertedType` as any sort of duplication. By the same logic, you would never use `GetTypeInfo` directly and always use this type inference thing in order to avoid duplication.

Reply:
>I don't really see pattern.GetTypeInfo().ConvertedType as any sort of duplication. 

I do.  But, as i mentioned, its not major for me.  

> By the same logic, you would never use GetTypeInfo directly and always use this type inference thing in order to avoid duplication.

I would be amenable to that.

Reply:
`{ X: Foo() } ` I don't understand what this means. I don't believe this code would ever be valid.

Reply:
my undertanding of ```{ X: Y }``` was that it was a subpattern checking that the field/property X matched the pattern 'Y'.  'Y' can be a constant pattern, allowing you to place a constant there, as such, you might want to generate the constant 'Y' if it does not already exist.  The type inference service helps determine the types of these things so that features like 'generate constant' etc. don't need to know about all these different cases.

Reply:
I feel like these 3 lines could be replaced with 1 using the actual compiler API. Can you please explain what sort of duplication you'd have and where exactly?

Reply:
> my undesrtanding of { X: Y } was that it was a subpattern checking that the field/property X matched the pattern 'Y'. 'Y' can be a constant pattern

1. I believe if we wanted to generate a property based on the type of the nested pattern, you could simply check the type of that pattern. Would that no work?
2. I'm not convinced this is something we should actually do. A constant pattern checks the type as well. Every time I've seen or written things like `x is 0`, x itself was not an int, but rather an object.

Reply:
> you might want to generate the constant 'Y' if it does not already exist

You could use type inference on the `Y` expression itself, not the constant pattern wrapping it. I still don't see any reason to change this API to not just work with expression anymore, especially considering the type safety we lose. Can you give an actual example where you want to use this on a pattern itself?

Reply:
Have added a test to demonstrate this.

Reply:
>I feel like these 3 lines could be replaced with just 1 using the actual compiler API. Can you please explain what sort of duplication you'd have and where exactly?

This code and hte type inference code would both have the special knowledge of having to go to the pattern and ask it for its converted type.  I would prefer the TypeInference code be the only one that has to know that's how to figure this type out.

Reply:
> A constant pattern checks the type as well. Every time I've seen or written things like x is 0, x itself was not an int, but rather an object.

IN the example test i've given, we want to know the type.  The constant we'd generate for Y would most likely want to be the same type as X's type for the user.

> Every time I've seen or written things like x is 0, x itself was not an int, but rather an object.

TypeInference in the IDE sense means just trying to infer a reasonable type.  So, if you have "x is Y", and 'x' is an object, we'll end up producing an object for 'Y', and that's ok.  But in other cases where you have a constant pattern, like ``{X: Y}```, it's totally reasonable to assume the type of Y is 'X' for hte purposes of generating a member.

Reply:
> You could use type inference on the Y expression itself, not the constant pattern wrapping it.

That's what we do.  And the Y expression says "i'm just a name, i should figure out what i mean in my parent context".  The parent context is the pattern, it says "i don't have enough information about msyelf to make a determination, so i'll defer to my parent context".  Eventually you hit a point where someone says "based on the code around me, here's  a reasonable type i think you should be."